src/provider/database.rs

```diff
@@ -6,21 +6,880 @@ use crate::{
 };
 use anyhow::{Context, Result};
 use bytes::Bytes;
-use futures::StreamExt;
+use futures::{Future, FutureExt, StreamExt};
+use serde::{Deserialize, Serialize};
 use std::{
-    collections::{BTreeSet, HashMap},
-    fmt, io,
+    collections::{BTreeMap, BTreeSet, HashMap},
+    fmt,
+    io::{self, Read},
     path::{Path, PathBuf},
+    pin::Pin,
     result,
     sync::{Arc, RwLock},
 };
-use tokio::sync::mpsc;
+use tokio::{io::AsyncReadExt, sync::mpsc};
+use tokio_util::either::Either;
+
+/// Data for a (possibly very large) blob
+///
+/// This can be used for both outboards and content. Be careful, this could
+/// point to a file that is a terabyte in size. So loading it into memory
+/// could be a problem.
+///
+/// Blob data can either be inline (tiny data, stored in the db) or in a file.
+/// The file can be in the default location or a custom location.
+#[derive(Clone, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub enum BlobData {
+    /// Data is stored inline in the db and exists in memory.
+    Inline(Bytes),
+    /// Data is stored in a file can be optionally cached in memory.
+    File(Option<PathBuf>),
+}
+
+impl fmt::Debug for BlobData {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Inline(data) => f.debug_tuple("Inline").field(&data.len()).finish(),
+            Self::File(path) => f.debug_tuple("File").field(path).finish(),
+        }
+    }
+}
+
+impl BlobData {
+    fn len(&self, default_path: impl Fn() -> PathBuf) -> io::Result<u64> {
+        Ok(match self {
+            Self::Inline(data) => data.len() as u64,
+            Self::File(custom_path) => {
+                let path = custom_path.as_ref().cloned().unwrap_or_else(default_path);
+                std::fs::metadata(path)?.len()
+            }
+        })
+    }
+
+    async fn len_async(&self, default_path: impl Fn() -> PathBuf) -> io::Result<u64> {
+        Ok(match self {
+            Self::Inline(data) => data.len() as u64,
+            Self::File(custom_path) => {
+                let path = custom_path.as_ref().cloned().unwrap_or_else(default_path);
+                tokio::fs::metadata(path).await?.len()
+            }
+        })
+    }
+
+    fn read_as_bytes(&self, default_path: impl Fn() -> PathBuf) -> io::Result<Bytes> {
```

Comment:
I'm used to seeing this as just `.bytes()` but maybe that's not a universal convention just something of the people I worked with.

Reply:
This actually does read from disk. A method that is called something like `.bytes()` for me implies that it is a very cheap conversion or just passing out something that already exists.

Reply:
It's mostly bikeshedding and I don't really mind, but `.bytes()` does not imply it's cheap, at least to me but it's all relative.  `.as_bytes()` would suggest it's cheap.  Maybe that's the reason I noticed this, because the current name includes that as a substring.